xchpst.8
'\" t
.Dd March 21, 2025
.Dt xchpst 8
.Os
.Sh NAME
.Nm xchpst
.Nd eXtended CHange Process STate
.Sh SYNOPSIS
.Nm
.Fl -help
.Nm
.Fl -version
.Nm
.Fl -exit Ns Op = Ns Ar retcode
.Nm
.Op Ar OPTIONS
.Op Fl -
.Pa command Ar ...
.Sh DESCRIPTION
The
.Nm
utility changes process state according to the supplied options and then
calls
.Fn exec
on a named executable with the positional arguments.
.Pp
.Nm
is a backwards-compatible extension to the
.Xr chpst 8
tool which is supplied with runit.
.Nm
enables runit service scripts to take advantage of hardening
capabilities available with recent Linux kernels such as namespaces
and capabilities.
.Nm
can set up shadow subtrees within the filesystem hierarchy to isolate
long-running services from parts of the system to which they ought to
need no access, e.g. with private
.Pa /tmp
areas and read-only
.Pa /usr.
.Ss Extended xchpst options
The extra options provided by
.Nm
are as follows:
.Bl -tag -width mmmbytesxx
.It Fl -help
Show help text and usage.
.It Fl -file Ar file
Read options from
.Ar file .
See
.Sx Options file
for the file format.
.It Fl -exit Ns Op = Ns Ar retcode
Exit immediately with exit status 0 if the given options are supported.
.Ar retcode
if specified.
.It Fl -mount-ns
Create new mount namespace.
Various other options also implicitly enable mount namespaces as this
is important to their operation; this option is rarely likely to be
needed to be specified explicitly.
.It Fl -net-ns
Create new network namespace.
This will more or less isolate the process from the networking subsystem.
.It Fl -uts-ns
Create new UTS namespace.
.It Fl -pid-ns
Create a PID namespace.
This implies
.Fl -fork-join
because a new process is needed to act as PID 1 and in order to be able
to mount a new procfs for the namespace.
.It Fl -fork-join
Fork a new process and wait for it to finish, passing on to the child
process any signals received by the
.Nm
process.
This option is necessary to take advantage of PID namespaces.
The exit status is that of the child process.
.It Fl -user-ns
Create a user namespace.
.It Fl -adopt-net Pa path
Adopt the network namespace bound to
.Pa path .
The binding will be deleted from the filesystem meaning that the
namespace will disappear when the process exits, if there is no other
reference to it. This allows the calling script to set up a suitable
networking environment for the process and hand it over.
.It Fl -new-root
Create a new root filesystem (will implicitly enable the creation
of a new mount namespace).
The new root filesystem is created as a tmpfs and all the top-level
directories in the original root filesystem are bind mounted and any
symlinks are replicated.
.It Fl -private-run
Mount an isolated
.Pa /run
directory for the process.
Unless
.Fl -new-root
is also specified, the old shared /run directory will still be accessible
if the stacked mount is removed.
.It Fl -private-tmp
Mount an isolated
.Pa /tmp
directory for the process.
Unless
.Fl -new-root
is also specified, the old shared /tmp directory will still be accessible
if the stacked mount is removed.
.It Fl -protect-home
Mount isolated
.Pa /home ,
.Pa /root and
.Pa /run/user
directories for the process.
Unless
.Fl -new-root
is also specified, the old shared host directories will still be accessible
if the stacked mounts are removed.
.It Fl -ro-sys
Convert
.Pa /usr
and
.Pa /boot
into read-only mounts.
Note that if the hardened process has the rights to unmount
filesystems, it can reveal the original writable filesystems.
The
.Fl -new-root
option is designed to prevent this.
Alternatively, use
.Fl -cap-bs-drop
to remove the
.Ql CAP_SYS_ADMIN
capability,
preventing the bind mount from being unmounted.
.It Fl -ro-home
Convert
.Pa /home ,
.Pa /root and
and
.Pa /run/usr
into read-only mounts.
This option has the same limitations as
.Fl -ro-sys .
.It Fl -ro-etc
Convert
.Pa /etc
into a read-only mount.
This option has the same limitations as
.Fl -ro-sys .
.It Fl -caps-bs-keep Ar capability Ns Op ,capability Ns ...
Keeps only the listed capabilities in the bounding set.
.It Fl -caps-bs-drop Ar capability Ns Op ,capability Ns ...
Drops the listed capabilities from the bounding set.
Use only one of the two options governing the bounding set.
.It Fl -caps-keep Ar capability Ns Op ,capability Ns ...
Retain the listed capabilities when dropping to a non-root
user.
.It Fl -caps-drop Ar capability Ns Op ,capability Ns ...
Drop the listed capabilities when dropping to a non-root
user, but retain all others.
.It Fl -no-new-privs
Prevent the target application from obtaining any new privileges.
See
.Xr PR_SET_NO_NEW_PRIVS 2const .
.It Fl -cpu-scheduler Ic other Ns | Ns Ic batch Ns | Ns Ic idle
Set the scheduler policy, as per
.Xr sched_setscheduler 2 .
.It Fl -io-scheduler Ic rt Ns | Ns Ic best-effort Ns | Ns Ic idle Ns Op : Ns Ar priority
Set the I/O scheduler policy and priority,
as per
.Xr ionice 1 .
.It Fl -cpus Ar start Ns Oo - Ns Ar end Ns Oo : Ns Ar stride Oc Oc Ns Op ,...
Set CPU affinity in the same format as
.Xr taskset 1 .
.It Fl -umask Ar mode
Set umask to the octal value
.Ar mode .
.It Fl -app Ar name
Override program name used for pre-creating system directories.
.It Fl -run-dir
Create a directory for the program under
.Pa /run ,
owned by the appropriate user.
.It Fl -state-dir
Create a directory for the program under
.Pa /var/lib ,
owned by the appropriate user.
.It Fl -log-dir
Create a directory for the program under
.Pa /var/log ,
owned by the appropriate user.
.It Fl -cache-dir
Create a directory for the program under
.Pa /var/cache ,
owned by the appropriate user.
.It Fl -login
Create a login environment, using the user specified by -u, -U or the current
user, in order of preference.
If this option is specified and no command is specified to be executed,
then the shell defined for the given user is launched, instead of an error
being returned.
.It Fl -oom Ar adjustment
Set the out-of-memory (OOM) score adjustment to
.Ar adjustment .
.It Fl s Ar bytes
Set soft limit for stack segment size.
.It Fl a Ar bytes
Set soft limit for address space size.
.It Fl -limit-memlock Ar bytes
Set soft limit for amount of locked memory.
.It Fl -limit-msgqueue Ar bytes
Set soft limit for message queue space for this user.
.It Fl -limit-nice Ar niceness
Set 20 minus the minimum niceness possible for this process.
.It Fl -limit-rtptio Ar prio
Set soft limit for real time priority of the process.
.It Fl -limit-rttime Ar ms
Set soft limit for amount of real time processing between blocking system calls.
.It Fl -limit-sigpending Ar number
Set soft limit for the number of pending signals permitted for the process.
.It Fl -limit-locks Ar number
Set soft limit for the number of file locks that this process may hold.
.It Fl -hardlimit
Also set the hard limit for any soft limit option that follows.
.It Fl @
Switch to chpst-compatible option handling only for the remaining
options. This is to support scripts that can convert an
.Nm
invocation into a command line for
.Nm chpst
if
.Nm
is not present on the system.
.El
.Ss chpst-compatible options
The options compatible with classic
.Nm chpst
are as follows:
.Bl -tag -width mmmbytes
.It Fl u Ar user Ns Oo Ar :group Ns Oc Ns ...
Set uid, gid and supplementary groups. Prepend the argument with a colon
for numerical inputs rather than names to be looked up. If no group is
specified then the specified user's group is used. There is no space
within the argument.
.It Fl U Ar user Ns Oo Ar :group Ns Oc
Like
.Fl u
but the environment variables
.Ev UID
and
.Ev GID
are set instead of changing

CHANGELOG
xchpst (0.6.2) UNRELEASED; urgency=medium

  * only drop user before unshare with user namespaces
  * avoid macros within tbl cells to fix mandoc formatting

 -- Andrew Bower <andrew@bower.uk>  Fri, 11 Apr 2025 23:20:53 +0100

xchpst (0.6.1) unstable; urgency=medium

  * add missing --limit-locks option
  * remove compat script example
  * tweak man page to avoid groff linting issues
  * add table comparing other tools' resource mapping options to man page

 -- Andrew Bower <andrew@bower.uk>  Fri, 21 Mar 2025 21:12:53 +0000

xchpst (0.6.0) unstable; urgency=medium

  * add --ro-home for read-only home directories
  * add --ro-etc for read-only config areas
  * add --hardlimit to augment chpst-style softlimit options
  * abort on failure to apply filesystem-related options
  * stop swamping terminal with full usage list on xchpst option error
  * extend man page
    - add table on option failure modes
    - add table with systemd option mappings

 -- Andrew Bower <andrew@bower.uk>  Fri, 21 Mar 2025 03:38:47 +0000

xchpst (0.5.0) unstable; urgency=medium

  * add --file to process an option file
  * add --oom for OOM score adjustment
  * extend rlimit options and handling
    - implement all currently-available rlimit options
    - add hard limit setting, delimited with ':'
    - accept '=' and 'infinity' as synonyms for 'unlimited'
  * normalise CPU and I/O scheduler option names
  * update man page, removing designation of interface as being provisional

 -- Andrew Bower <andrew@bower.uk>  Thu, 13 Mar 2025 21:28:03 +0000

xchpst (0.4.1) unstable; urgency=medium

  * place roots under subdirectory with fallback alternatives for /run
  * restore dropping user before unshare

 -- Andrew Bower <andrew@bower.uk>  Fri, 14 Feb 2025 23:54:20 +0000

xchpst (0.4.0) unstable; urgency=medium

  * Add directory pre-creation options
    - add --run-dir
    - add --state-dir
    - add --cache-dir
    - add --log-dir
    - add --app option to override pre-created directory name
  * Add --umask option
  * Add --login to set login environment and run shell if no command specified
  * Fix tracking of capabilities being present
  * Return options error 100 with --exit when appropriate.
    (Creates an option check mode)
  * Create necessary mount ns with pid ns
  * Move new roots to under /run/xchpst instead of /tmp
  * When running in chpst-emulation mode, only accept chpst options

 -- Andrew Bower <andrew@bower.uk>  Wed, 12 Feb 2025 20:19:56 +0000

xchpst (0.3.0) unstable; urgency=medium

  * Add new options
    - add --cpus
    - add --io-nice
    - add --uts-ns
  * Restore pre-C23 compat for gcc-12

 -- Andrew Bower <andrew@bower.uk>  Wed, 29 Jan 2025 21:31:36 +0000

xchpst (0.2.1) unstable; urgency=medium

  * Enable -Wextra and build cleanly therewith
  * Demote xchpst.fake to subdirectory, not installed

 -- Andrew Bower <andrew@bower.uk>  Tue, 07 Jan 2025 17:30:01 +0000

xchpst (0.2.0) unstable; urgency=medium

  * Add new options
    - add --no-new-privs
    - add --scheduler
    - add --protect-home
  * Provide xchpst.fake example to enable runscripts to use xchpst
    features when present and chpst features otherwise.
  * Fix minor issues

 -- Andrew Bower <andrew@bower.uk>  Sun, 05 Jan 2025 23:28:09 +0000

xchpst (0.1.0) unstable; urgency=medium

  * First release of xchpst (eXtended CHange Process STate) tool.
  * This is a freshly-written tool, backwards-compatible with chpst.
  * Versions numbered 0.x have an unstable user interface.
  * See the man page, xchpst (8), for documentation.

 -- Andrew Bower <andrew@bower.uk>  Sun, 29 Dec 2024 23:09:17 +0000


xchpst.c
/* SPDX-License-Identifier: MIT */
/* SPDX-FileCopyrightText: (c) Copyright 2024,2025 Andrew Bower <andrew@bower.uk> */

/* xchpst: eXtended Change Process State
 * A tool that is backwards compatible with chpst(8) from runit(8),
 * offering additional options to harden process with namespace isolation
 * and more. */

#include <assert.h>
#include <ctype.h>
#include <fcntl.h>
#include <getopt.h>
#include <grp.h>
#include <errno.h>
#include <poll.h>
#include <sched.h>
#include <signal.h>
#include <stdarg.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <linux/prctl.h>
#include <linux/ioprio.h>
#include <sys/file.h>
#include <sys/dir.h>
#include <sys/mount.h>
#include <sys/stat.h>
#include <sys/pidfd.h>
#include <sys/prctl.h>
#include <sys/signalfd.h>
#include <sys/time.h>
#include <sys/wait.h>
#include <sys/syscall.h>

#include "xchpst.h"
#include "caps.h"
#include "join.h"
#include "env.h"
#include "options.h"
#include "rootfs.h"
#include "mount.h"
#include "precreate.h"

static const char *version_str = STRINGIFY(PROG_VERSION);
#ifdef PROG_DEFAULT
static const char *default_app = STRINGIFY(PROG_DEFAULT);
#else
static const char *default_app = "xchpst";
#endif
#define PROG "xchpst"

const struct app apps[] = {
  { COMPAT_CHPST,     "chpst",     .long_opts = false },
  { COMPAT_XCHPST,    "xchpst",    .long_opts = true },
  { COMPAT_SOFTLIMIT, "softlimit", .long_opts = false },
  { COMPAT_ENVDIR,    "envdir",    false, 1, { OPT_ENVDIR } },
  { COMPAT_PGRPHACK,  "pgrphack",  false, 1, { OPT_PGRPHACK } },
  { COMPAT_SETUIDGID, "setuidgid", false, 1, { OPT_SETUIDGID } },
  { COMPAT_ENVUIDGID, "envuidgid", false, 1, { OPT_ENVUIDGID } },
  { COMPAT_SETLOCK,   "setlock",   false, 1, { OPT_LOCK_WAIT } },
};
#define max_apps ((ssize_t) (sizeof apps / (sizeof *apps)))

struct runtime runtime = { 0 };

static const char *std_run_dir = "/run/" PROG;
static const char *fallback_run_dir = "/tmp/run-" PROG;
static int run_dir_fd = -1;
char *run_dir;

static struct {
  char uid[16];
  char gid[16];
} extra_env;

static void version(FILE *out) {
  fprintf(out,
          "xchpst-%s (c) Copyright 2024,2025 Andrew Bower <andrew@bower.uk>\n",
          version_str);
}

static void usage(FILE *out) {
  version(out);
  fprintf(out, "\nusage: %s OPTIONS [--]",
          program_invocation_short_name);
  options_print_positional(out);
  fprintf(out, " PROG...    launch PROG with changed process state\n");
  options_explain_positional(out);
  options_print(out);
}

int ensure_dir(int dirfd, const char *path, int *fd, mode_t mode) {
  for (int creating = 0; *fd == -1 && creating <= 1; creating++) {
    *fd = openat(dirfd, path, O_DIRECTORY | O_CLOEXEC);
    if (*fd == -1 && !creating && (errno == ENOENT))
      mkdirat(dirfd, path, mode);
  }
  if (*fd == -1 && errno != EEXIST)
    return -1;
  return 0;
}

int get_run_dir(void) {
  int rc = ensure_dir(-1, std_run_dir, &run_dir_fd, 0700);
  if (rc == -1) {
    char *xdg_run_dir = getenv("XDG_RUNTIME_DIR");
    if (xdg_run_dir) {
      char *path;
      rc = asprintf(&path, "%s/%s", xdg_run_dir, PROG);
      if (rc != -1) {
        rc = ensure_dir(-1, path, &run_dir_fd, 0700);
        if (rc != -1) {
          run_dir = path;
        } else {
          free(path);
        }
      }
    }
  } else {
    run_dir = strdup(std_run_dir);
  }
  if (rc == -1 &&
      ensure_dir(-1, fallback_run_dir, &run_dir_fd, 0700) != -1)
    run_dir = strdup(fallback_run_dir);

  return run_dir_fd;
}

static int write_once(const char *file, const char *fmt, ...) {
  int fd = open(file, O_WRONLY);
  char *text;
  ssize_t len;
  int rc = 0;
  va_list args;

  if ((rc = fd == -1 ? 1 : 0))
    goto fail0;

  va_start(args, fmt);
  len = vasprintf(&text, fmt, args);
  va_end(args);
  if (text == NULL || len == -1) {
    rc = -1;
    goto fail;
  }

  rc = write(fd, text, len);
  rc = (rc == len ? 0 : -1);

  free(text);

fail:
  close(fd);
fail0:
  if (rc != 0)
    fprintf(stderr, "writing to %s: %s\n", file, strerror(errno));
  return rc;
}

void set_rlimit(int resource, struct limit *option) {
  struct rlimit prev;

  if (option->soft_specified || option->hard_specified) {
    if (getrlimit(resource, &prev) != 0) {
      fprintf(stderr, "warning: resource type %d cannot be controlled on this kernel\n", resource);
      return;
    }
  } else {
    return;
  }

  if (option->hard_specified)
    prev.rlim_max = option->limits.rlim_max;

  if (option->soft_specified) {
    if (option->limits.rlim_cur != RLIM_INFINITY &&
        option->limits.rlim_max != RLIM_INFINITY &&
        option->limits.rlim_cur > prev.rlim_max) {
      fprintf(stderr, "warning capping requested %d soft limit from %lld to maximum %lld\n",
              resource, (long long) option->limits.rlim_cur, (long long) prev.rlim_max);
      prev.rlim_cur = prev.rlim_max;
    } else {
      prev.rlim_cur = option->limits.rlim_cur;
    }
  }

  if (setrlimit(resource, &prev) != 0) {
    fprintf(stderr, "warning: failed to set type %d soft limit\n", resource);
  }
}

void set_resource_limits(void) {
  set_rlimit(RLIMIT_DATA, &opt.rlimit_data);
  set_rlimit(RLIMIT_AS, &opt.rlimit_as);
  set_rlimit(RLIMIT_STACK, &opt.rlimit_stack);
  set_rlimit(RLIMIT_MEMLOCK, &opt.rlimit_memlock);
  set_rlimit(RLIMIT_RSS, &opt.rlimit_rss);
  set_rlimit(RLIMIT_NOFILE, &opt.rlimit_nofile);
  set_rlimit(RLIMIT_NPROC, &opt.rlimit_nproc);
  set_rlimit(RLIMIT_FSIZE, &opt.rlimit_fsize);
  set_rlimit(RLIMIT_CORE, &opt.rlimit_core);
  set_rlimit(RLIMIT_CPU, &opt.rlimit_cpu);
  set_rlimit(RLIMIT_MSGQUEUE, &opt.rlimit_msgqueue);
  set_rlimit(RLIMIT_NICE, &opt.rlimit_nice);
  set_rlimit(RLIMIT_RTPRIO, &opt.rlimit_rtprio);
  set_rlimit(RLIMIT_RTTIME, &opt.rlimit_rttime);
  set_rlimit(RLIMIT_SIGPENDING, &opt.rlimit_sigpending);
  set_rlimit(RLIMIT_LOCKS, &opt.rlimit_locks);
}

bool drop_user(uid_t uid, gid_t gid) {
  int rc;
  int i;

  /* You do this backwards: supplemental groups first */
  gid_t *groups = malloc(sizeof(gid_t) * opt.users_groups.num_supplemental);

  if (opt.caps_op != CAP_OP_NONE) {
    /* Postpone the loss of capabilities until after the user switch
     * and drop them when ready. */
    rc = prctl(PR_SET_KEEPCAPS, 1);
  }

  if (!groups)
    return false;

  for (i = 0; i < opt.users_groups.num_supplemental; i++)
    groups[i] = opt.users_groups.supplemental[i].gid;
  rc = (setgroups(i, groups) == -1 ? errno : 0);
  free(groups);
  if (rc) {
    perror("setgroups");
    return false;
  }

  /* Then main group */
  rc = setresgid(gid, gid, gid);
  if (rc) {
    perror("setresgid");
    return false;
  }

  /* Then the actual user */
  rc = setresuid(uid, uid, uid);
  if (rc) {
    perror("setresgid");
    return false;
  }
  return true;
}

static const struct app *find_app(const char *name) {
  const struct app *app;
  const char *ext = strchrnul(name, '.');

  for (app = apps;
       app - apps < max_apps && strncmp(name, app->name, ext - name);
       app++);

  return app;
}

int main(int argc, char *argv[]) {
  sigset_t newmask;
  sigset_t oldmask;
  char **sub_argv;
  char *executable;
  char *new_root = NULL;
  char *old_root = NULL;
  int sub_argc;
  pid_t child;
  int optind;
  int rc = 0;
  int ret = CHPST_ERROR_CHANGING_STATE;
  int lock_fd = -1;
  bool in_new_root = false;
  uid_t uid;
  gid_t gid;
  int fd;

  /* As which application were we invoked? */
  opt.app = find_app(program_invocation_short_name);
  if (opt.app - apps == max_apps)
    opt.app = find_app(default_app);
  assert(opt.app - apps != max_apps);

  if (!options_init())
    return CHPST_ERROR_OPTIONS;
  optind = options_parse(argc, argv);

  if (is_verbose())
    fprintf(stderr, "invoked as %s(%s)\n", opt.app->name, program_invocation_short_name);

  if (!set(OPT_FORK_JOIN) &&
      (opt.new_ns & CLONE_NEWPID)) {
    if (is_verbose())
      fprintf(stderr, "also going to do fork-join since new PID namespace requested\n");
    enable(OPT_FORK_JOIN);
  }

  if (!(opt.new_ns & CLONE_NEWNS) &&
      (set(OPT_NET_NS) || set(OPT_PRIVATE_RUN) || set(OPT_PRIVATE_TMP) ||
       set(OPT_RO_SYS) || set(OPT_RO_HOME) || set(OPT_RO_ETC) ||
       set(OPT_NEW_ROOT) || set(OPT_PID_NS))) {
    if (is_verbose())
      fprintf(stderr, "also creating mount namespace implicitly due to other options\n");
    opt.new_ns |= CLONE_NEWNS;
  }

  if (opt.exit) {
    ret = opt.error ? CHPST_ERROR_OPTIONS : opt.retcode;
    goto finish0;
  }

  if (opt.help)
    usage(stdout);
  else if (opt.version)
    version(stdout);

  if (opt.help || opt.version) {
    ret = CHPST_OK;
    goto finish0;
  }

  if (optind == argc && !set(OPT_LOGIN))
    opt.error = true;

  if (opt.error) {
    const struct option_info *help_option = find_option(OPT_HELP, NULL);
    if (help_option && opt.app->long_opts && help_option->long_name)
      fprintf(stderr, "%s: error in options. Run %s --%s for usage\n",
              program_invocation_short_name,
              program_invocation_short_name,
              help_option->long_name);
    else
      usage(stderr);
    ret = CHPST_ERROR_OPTIONS;
    goto finish0;
  }

  /* Do xchpsty-type things now! */
  sub_argc = argc - optind;
  if (sub_argc == 0)
    sub_argc++;
  sub_argv = malloc((sub_argc + 1) * sizeof *sub_argv);
  memcpy(sub_argv, argv + optind, (argc - optind) * sizeof *sub_argv);
  sub_argv[sub_argc] = NULL;

  if (set(OPT_UMASK))
    umask(opt.umask);

  if (set(OPT_OOM))
    write_once("/proc/self/oom_score_adj", "%ld", opt.oom_adjust);

  if (opt.lock_file) {
    if (opt.lock_nowait_override)
      opt.lock_wait = false;
    lock_fd = open(opt.lock_file, O_WRONLY | O_NDELAY | O_APPEND | O_CREAT, 0600);
    if (lock_fd != -1 && flock(lock_fd, LOCK_EX | (opt.lock_wait ? 0 : LOCK_NB)) == -1) {
      close(lock_fd);
      lock_fd = -1;
    }
    if (lock_fd == -1) {
      if (opt.lock_quiet)
        ret = CHPST_ERROR_EXIT;
      else
        fprintf(stderr, "error obtaining lock, %s\n", strerror(errno));
      goto finish;
    }
  }

  if (set(OPT_PGRPHACK)) {
    rc = setsid();
    if (rc == -1) {
      perror("setsid");
      goto finish;
    } else if (is_verbose()) {
      fprintf(stderr, "new session id: %d\n", rc);
    }
  }

  if (opt.env_dir && !read_env_dir(opt.env_dir))
    goto finish;

  if (set(OPT_ENVUIDGID)) {
    struct sys_entry *entry;
    if ((entry = &opt.env_users_groups.user)->resolved) {
      rc = snprintf(extra_env.uid, sizeof extra_env.uid, "UID=%d", entry->uid);
      if (rc == -1 || rc == sizeof extra_env.uid) {
        perror("creating UID env");
        goto finish;
      }
      putenv(extra_env.uid);
    }
    if ((entry = &opt.env_users_groups.group)->resolved) {
      rc = snprintf(extra_env.gid, sizeof extra_env.gid, "GID=%d", entry->gid);
      if (rc == -1 || rc == sizeof extra_env.uid) {
        perror("creating UID env");
        goto finish;
      }
      putenv(extra_env.gid);
    }
  }

  if (set(OPT_NICE)) {
    int newnice;

    errno = 0;
    newnice = nice(opt.niceness);
    if (errno) {
      fprintf(stderr, "could not change niceness, %s\n", strerror(errno));
      goto finish;
    }
    if (is_verbose()) {
      fprintf(stderr, "now at niceness %d\n", newnice);
    }
  }

  if (set(OPT_IO_SCHED)) {
    if (syscall(SYS_ioprio_set,IOPRIO_WHO_PROCESS, 0, opt.ionice_prio) == -1) {
      fprintf(stderr, "warning: failed to set I/O scheduling class\n");
    } else if (is_verbose()) {
      fprintf(stderr, "set IO class to %d:%ld\n",
              IOPRIO_PRIO_CLASS(opt.ionice_prio),
              IOPRIO_PRIO_DATA(opt.ionice_prio));
    }
  }

  if (opt.cpu_affinity.size &&
      sched_setaffinity(0, opt.cpu_affinity.size,
                        opt.cpu_affinity.mask) == -1)
    perror("could not set CPU affinity");

  if (set(OPT_CPU_SCHED) &&
      sched_setscheduler(0, opt.sched_policy, &((struct sched_param) {})) == -1)
    perror("could not change scheduler policy");

  if ((opt.cap_bounds_op != CAP_OP_NONE ||
       opt.caps_op != CAP_OP_NONE) &&
      runtime.absent.caps) {
    fprintf(stderr, "ignoring capabilities as not supported on system");
    opt.cap_bounds_op = opt.caps_op = CAP_OP_NONE;
  }

  if (opt.cap_bounds_op != CAP_OP_NONE)
    if (!set_capabilities_bounding_set())
      goto finish;

  if (set(OPT_SETUIDGID) && opt.users_groups.user.resolved) {
    gid = opt.users_groups.group.gid;
    uid = opt.users_groups.user.uid;
  } else {
    uid = getuid();
    gid = getgid();
  }

  /* Set a login environment */
  if (set(OPT_LOGIN)) {
    struct users_groups *ug = NULL;
    struct users_groups current = { 0 };

    /* Prefer -u, -U then current user, but if the preferred choice
     * is not resolved then we don't set a login environment, rather
     * than picking an unpredictable one. */
    if (set(OPT_SETUIDGID))
      ug = &opt.users_groups;
    else if (set(OPT_ENVUIDGID))
      ug = &opt.env_users_groups;
    else
      usrgrp_resolve_uid(ug = &current, getuid());

    if (ug->user.resolved) {
      if (ug->username && *ug->username) {
        setenv("USER", ug->username, 1);
        setenv("LOGNAME", ug->username, 1);
      }
      if (ug->home && *ug->home) {
        setenv("HOME", ug->home, 1);
      }
      if (ug->shell && *ug->shell) {
        setenv("SHELL", ug->shell, 1);
      }
    }
  }

  if (argc == optind)
    sub_argv[0] = getenv("SHELL");

  executable = sub_argv[0];
  if (opt.argv0)
    sub_argv[0] = opt.argv0;

  if (opt.app_name == NULL)
    opt.app_name = basename(sub_argv[0]);

  {
    uid_t o = set(OPT_SETUIDGID) ? uid : (uid_t) -1;
    gid_t g = set(OPT_SETUIDGID) ? gid : (gid_t) -1;

    if (set(OPT_RUN_DIR) &&
        precreate_dir("/run", 0755, o, g) == -1)
      goto finish;

    if (set(OPT_STATE_DIR) &&
        precreate_dir("/var/lib", 0755, o, g) == -1)
      goto finish;

    if (set(OPT_CACHE_DIR) &&
        precreate_dir("/var/cache", 0755, o, g) == -1)
      goto finish;

    if (set(OPT_LOG_DIR) &&
        precreate_dir("/var/log", 0755, o, g) == -1)
      goto finish;
  }

  /* Iff using a user namespace, drop the user first */
  if (set(OPT_SETUIDGID) &&
      opt.new_ns & CLONE_NEWUSER &&
      opt.users_groups.user.resolved &&
      !drop_user(uid, gid))
      goto finish;

  if (opt.new_ns) {
    rc = unshare(opt.new_ns);
    if (rc == -1) {
      perror(NAME_STR ": unshare()");
      goto finish;
    }
    if (opt.verbosity > 0) fprintf(stderr, "created 0xb%b namespaces\n", opt.new_ns);

    if (opt.new_ns & CLONE_NEWNS) {
      rc = mount(NULL, "/", NULL,
                 MS_REC | MS_SLAVE, NULL);
      if (rc == -1)
        fprintf(stderr, "recursive remounting / as MS_SLAVE: %s", strerror(errno));
    }

    if (opt.new_ns & CLONE_NEWNET)
      special_mount("/sys", "sysfs", "sysfs", NULL);
  }

  if (opt.new_ns & CLONE_NEWUSER) {
    rc = prctl(PR_SET_DUMPABLE, 1);
    setgroups(0, NULL);
    write_once("/proc/self/setgroups", "%s", "deny\n");
    write_once("/proc/self/gid_map", "%u %u %u\n", 0, gid, 1);
    write_once("/proc/self/uid_map", "%u %u %u\n", 0, uid, 1);
    if (setresgid(0, 0, 0) != 0 ||
        setresuid(0, 0, 0) != 0)
      fprintf(stderr, "warning: error becoming root in user namespace, %s\n", strerror(errno));
  }

  if (opt.net_adopt) {
    const char *failed_op = NULL;
    if ((fd = open(opt.net_adopt, O_RDONLY)) != -1) {
      if ((rc = setns(fd, CLONE_NEWNET)) == 0) {
        if ((rc = umount2(opt.net_adopt, MNT_DETACH)) == 0) {
          if ((rc = unlink(opt.net_adopt)) == -1)
            failed_op = "unlink";
        } else failed_op = "umount2";
      } else {
        failed_op = "setns"; close(fd);
      }
    } else failed_op = "open";
    if (failed_op) {
      perror(failed_op);
      goto finish;
    }
    if (opt.verbosity > 0) fprintf(stderr, "adopted net ns\n");
  }

  if (set(OPT_NEW_ROOT)) {
    if (!create_new_root(basename(executable), &new_root, &old_root))
      goto finish;
  }

  set_resource_limits();

  if (set(OPT_FORK_JOIN)) {
    /* Save old signal mask for re-use by child and block all signals
     * in the parent so we can get them delivered by signalfd. */
    sigfillset(&newmask);
    sigdelset(&newmask, SIGCHLD);
    sigdelset(&newmask, SIGBUS);
    sigdelset(&newmask, SIGFPE);
    sigdelset(&newmask, SIGILL);
    sigdelset(&newmask, SIGSEGV);
    rc = sigprocmask(SIG_SETMASK, &newmask, &oldmask);
    if (rc == -1) {
      perror("setting up mask for signalfds");
      goto finish;
    }

    child = fork();
    if (child == -1) {
      perror("fork");
      goto finish;
    } else if (child != 0) {
      goto join;
    } else {
      if (sigprocmask(SIG_SETMASK, &oldmask, NULL) == -1)
        perror("warning: could not restore signal mask in child");
    }
  }

  /*************************************
   *  Inside child if fork-join used   *
   *************************************/

  if (set(OPT_NEW_ROOT)) {
    if (!pivot_to_new_root(new_root, old_root))
      goto finish;
    else
      in_new_root = true;
  }

  if (opt.chroot) {
    rc = chdir(opt.chroot);
    if (rc == -1) {
      perror("chdir for chroot");
      goto finish;
    }
    rc = chroot(".");
    if (rc == -1) {
      perror("chroot");
      goto finish;
    }
    if (is_verbose())
      fprintf(stderr, "entered chroot: %s\n", opt.chroot);
  }

  if (opt.chdir) {
    rc = chdir(opt.chdir);
    if (rc == -1) {
      perror("chdir");
      goto finish;
    }
    if (is_verbose())
      fprintf(stderr, "change directory: %s\n", opt.chdir);
  }

  if (opt.new_ns & CLONE_NEWPID) {
    special_mount("/proc", "proc", "procfs", NULL);
  }

  if (set(OPT_PRIVATE_RUN) &&
      private_mount("/run") == -1)
    goto finish;

  if (set(OPT_PRIVATE_TMP) &&
      (private_mount("/tmp") == -1 ||
       private_mount("/var/tmp") == -1))
    goto finish;

  if (set(OPT_PROTECT_HOME) &&
      (private_mount("/home") == -1 ||
       private_mount("/root") == -1 ||
       private_mount("/run/user") == -1))
    goto finish;

  if (!set(OPT_PROTECT_HOME) && set(OPT_RO_HOME) &&
      (remount_ro("/home") == -1 ||
       remount_ro("/root") == -1 ||
       remount_ro("/run/user") == -1))
    goto finish;

  if (set(OPT_RO_SYS) &&
      remount_sys_ro() == -1)
    goto finish;

  if (set(OPT_RO_ETC) &&
      (remount_ro("/etc") == -1))
    goto finish;

  if (set(OPT_SETUIDGID) &&
      (opt.new_ns & CLONE_NEWUSER) == 0 &&
      opt.users_groups.user.resolved &&
      !drop_user(uid, gid))
      goto finish;

  if (opt.caps_op != CAP_OP_NONE)
    if (!drop_capabilities())
      goto finish;

  for (unsigned int close_fds = opt.close_fds; close_fds; close_fds &= ~(1 << fd))
    close(fd = /*stdc_trailing_zeros*/ __builtin_ctz(close_fds));

  if (set(OPT_NO_NEW_PRIVS) && prctl(PR_SET_NO_NEW_PRIVS, 1L, 0L, 0L, 0L) == -1)
    perror("could not honour --no-new-privs");

  /* Launch the target */
  rc = execvp(executable, sub_argv);

  /* Handle errors launching */
  assert(rc == -1);
  perror(NAME_STR ": execvp");

join:
  if (set(OPT_FORK_JOIN) && child != 0)
    join(child, &newmask, &oldmask, &ret);

finish:
  /* Actions here should be
     1) suitable for if exec() fails.
     2) clean up if --fork-join is used.
     3) not be necessary when --fork-join is not used.
   */

  if (new_root && !in_new_root) {
    if (umount2(new_root, MNT_DETACH) == -1)
      fprintf(stderr, "umount2(%s): %s\n", new_root, strerror(errno));
    if (rmdir(new_root) == -1)
      fprintf(stderr, "rmdir(%s): %s\n", new_root, strerror(errno));
    free_rootfs_data();
  }

  if (run_dir_fd != -1)
    close(run_dir_fd);

  if (lock_fd != -1)
    close(lock_fd);

  free(new_root);
  free(old_root);
  free(sub_argv);
  free(run_dir);

finish0:
  options_free();
  return ret;
}
